### DISCUSSION TITLE
Use perf_client on models with string as input and variable shape

### DISCUSSION THREAD

**Discussion by ricard-borras-veriff** [triton-inference-server/server#5414] [Q&A] [answered]
I have a question about perf_client with ensemble_models: I have an ensemble model with this input config:

input [
  {
    name: "INPUT"
    data_type: TYPE_STRING
    dims: [ -1 ]
  }
]

This model works properly when sending images with different sizes as encoded strings. How can I use perf_client on this? If I try

`perf_client -m face_detection_pipeline`

I get an error:

`error: Failed to init manager inputs: input INPUT contains dynamic shape, provide shapes to send along with the request`

 If I set the shape parameter: 

`perf_client -m face_detection_pipeline --shape INPUT:320,240,3`

It complains: 

`Thread [0] had error: [request id: <id_unknown>] unexpected shape for input 'INPUT' for model 'face_detection_pipeline'. Expected [-1], got [320,240,3]`

How to proceed with this?

I am using perf_client from docker image tritonserver:22.12-py3-sdk 

Thanks!

**Comment by the-david-oy** [comment, marked as answer]
You have one variable dimension, but you're trying to set three. Your config either needs to have the input be [-1, -1, -1] or the shape needs to be one-dimensional, like `--shape INPUT:320`.

**Comment by ricard-borras-veriff** [comment]
Great! It works now by using --shape INPUT:320. However, models complains about data decoding, I suppose because that send data is not a valid image.  How should I send a test image using perf_client when input is with strings? 
Thanks!

**Comment by the-david-oy** [comment]
Depends on your model. What is it expecting? The way you wrote your config, it should have only one dimension. If, for example, your model expects the above dimensions, then you could change the INPUT dimensions in the config to "[320,240,3]" and run perf_client with your original `--shape INPUT:320,240,3`.

**Reply by ricard-borras-veriff** [reply]
My model is expecting variable shape images, the only restriction I can set is that they are 3-channeled. However, if I set input to [-1,-1,3] I will have the same problem with perf_client as with -1, correct?
I suppose that my original question could be rephrased as: "How could I send a given input image with perf_client to a model expecting variable input sizes such as [-1] or [-1,-1,-3]?
Thanks!

**Reply by the-david-oy** [reply]
You'd set the config to [-1, -1, 3] and you'd add a flag to perf_client: `--shape INPUT:320,240,3`.

**Comment by ricard-borras-veriff** [comment]
Yes, that would work for the shape, but what about the content itself? How
to send the image pixels?

**Reply by the-david-oy** [reply]
If you need specific data needs, then you can follow the documentation for perf analyzer [here](https://github.com/triton-inference-server/client/blob/main/src/c++/perf_analyzer/README.md#input-data).

You can set data from an image, pre-defined values, or a directory. The default is random, but there are many ways to set your own data (or even use all zeroes, if that works for your use case).

**Comment by ricard-borras-veriff** [comment]
Thanks for the doc link, there should be the answer for sure!
